File: Problems/0100_0199/0107_Binary_Tree_Level_Order_Traversal2/Project_CS/Binary_Tree_Level_Order_Traversal2.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public List<List<int>> LevelOrderBottom(TreeNode root)
    {
        var result = new List<List<int>>();
        if(root == null) return result;
        
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        
        while(true)
        {
            int nodeCount = queue.Count;
            if(nodeCount == 0)
            {
                break;
            }
            
            var subList = new List<int>();
            while(nodeCount > 0)
            {
                var dataNode = (TreeNode)queue.Dequeue();
                
                subList.Add(dataNode.val);
                
                if(dataNode.left != null)
                {
                    queue.Enqueue(dataNode.left);
                }
                
                if(dataNode.right != null)
                {
                    queue.Enqueue(dataNode.right);
                }
                
                nodeCount--;
            }
            
            result.Insert(0, subList);
        }
        
        return result;
    }

    public string ListListArrayToString(List<List<int>> list)
    {
        if (list == null)
            return "";

        string resultStr = "[\n";
        for (int i = 0; i < list.Count; ++i)
        {
            if (list[i].Count <= 0)
                continue;

            resultStr += "  [" + String.Join(",", list[i]) + "]\n";
        }

        return resultStr + "]\n";
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string flds = args.Replace("\"", "").Replace("[", "").Replace("]", "").Trim();

        OperateTreeNode ope_t = new OperateTreeNode();
        TreeNode root = ope_t.CreateTreeNode(flds);

        Console.Write("root = \n" + ope_t.TreeToStaircaseString(root));
        Console.WriteLine("root = " + ope_t.Tree2str(root));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        List<List<int>> result = LevelOrderBottom(root);

        sw.Stop();
        Console.WriteLine("result = \n" + ListListArrayToString(result));
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0100_0199/0107_Binary_Tree_Level_Order_Traversal2/Project_Python3/Binary_Tree_Level_Order_Traversal2.py
# coding: utf-8

from collections import deque
import os
import sys
import time

from TreeNode.Codec import Codec
from TreeNode.TreeNode import TreeNode
from TreeNode.OperateTreeNode import OperateTreeNode

class Solution:
    def levelOrderBottom(self, root: 'TreeNode') -> 'List[List[int]]':
        if not root:
            return []
        ans=[[root.val]]
        queue=deque([root])
        while queue:
            cur_val_list=[]
            for i in range(len(queue)):
                root=queue.popleft()
                if root.left:
                    cur_val_list.append(root.left.val)
                    queue.append(root.left)
                if root.right:
                    cur_val_list.append(root.right.val)
                    queue.append(root.right)
            if cur_val_list:
                ans.insert(0,cur_val_list)
        return ans

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip()

    ope_t = OperateTreeNode()
    if len(flds) > 0:
        root = ope_t.createTreeNode(flds)
    else:
        root = None

    print("root = \n{0}".format(ope_t.treeToStaircaseString(root)))
    print("root = {0}".format(ope_t.tree2str(root)))

    sl = Solution()
    time0 = time.time()

    result = sl.levelOrderBottom(root)

    time1 = time.time()

    print("result = {0}".format(result))
    print("Execute time ... : %f[s]\n" %(time1 - time0))

if __name__ == "__main__":
    main()
